devices/uart-nrf9160/uart.c

```diff
@@ -0,0 +1,367 @@
+	*(uart->base + uarte_starttx) = 1u;
+	while (*(uart->base + uarte_events_txstarted) != 1u) {
+		;
+	}
+
+	*(uart->base + uarte_events_txstarted) = 0u;
+
+	while (*(uart->base + uarte_events_endtx) != 1u) {
+		;
+	}
+
+	*(uart->base + uarte_events_endtx) = 0u;
+
+	return (size_t)(*(uart->base + uarte_txd_amount));
+}
+
+
+static ssize_t uart_write(unsigned int minor, addr_t offs, const void *buff, size_t len)
+{
+	uart_t *uart;
+	size_t cnt;
+
+	if ((uart = uart_getInstance(minor)) == NULL) {
+		return -EINVAL;
+	}
+
+	hal_memcpy((void *)uart->tx_dma, buff, len);
+
+	cnt = uart_send(uart, len);
+
+	return (ssize_t)cnt;
+}
+
+
+static int uart_sync(unsigned int minor)
+{
+	uart_t *uart;
+
+	if ((uart = uart_getInstance(minor)) == NULL)
+		return -EINVAL;
+
+	while (!lib_cbufEmpty(&uart->cbuffTx)) {
+		;
+	}
+
+	/* endtx flag is asserted in uart_send so mustn't be checked here */
+
+	return EOK;
+}
+
+
+static int uart_done(unsigned int minor)
+{
+	uart_t *uart;
+
+	if ((uart = uart_getInstance(minor)) == NULL) {
+		return -EINVAL;
+	}
+
+	/* Wait for transmission activity complete */
+	(void)uart_sync(minor);
+
+	*(uart->base + uarte_stoptx) = 1u;
+	*(uart->base + uarte_stoprx) = 1u;
+	/* disable all uart interrupts */
+	*(uart->base + uarte_intenclr) = 0xffffffffu;
+	hal_cpuDataMemoryBarrier();
+	*(uart->base + uarte_enable) = 0u;
+	hal_cpuDataMemoryBarrier();
+
+	hal_interruptsSet(uartInfo[minor].irq, NULL, NULL);
+
+	return EOK;
+}
+
+
+static int uart_map(unsigned int minor, addr_t addr, size_t sz, int mode, addr_t memaddr, size_t memsz, int memmode, addr_t *a)
+{
+	if (minor >= UART_MAX_CNT) {
+		return -EINVAL;
+	}
+
+	/* Device mode cannot be higher than map mode to copy data */
+	if ((mode & memmode) != mode) {
+		return -EINVAL;
+	}
+
+	/* uart is not mappable to any region */
+	return dev_isNotMappable;
+}
+
+
+/* Init pins according to nrf9160 product specification */
+static int uart_configPins(unsigned int minor)
+{
+	int ret;
+
+	if ((ret = _nrf91_gpioConfig(uartInfo[minor].txpin, gpio_output, gpio_nopull)) < 0) {
+		return ret;
+	}
+
+	if ((ret = _nrf91_gpioConfig(uartInfo[minor].rxpin, gpio_input, gpio_nopull)) < 0) {
+		return ret;
+	}
+
+	if ((ret = _nrf91_gpioConfig(uartInfo[minor].rtspin, gpio_output, gpio_nopull)) < 0) {
+		return ret;
+	}
+
+	if ((ret = _nrf91_gpioConfig(uartInfo[minor].ctspin, gpio_input, gpio_pulldown)) < 0) {
+		return ret;
+	}
+
+	if ((ret = _nrf91_gpioSet(uartInfo[minor].txpin, gpio_high)) < 0) {
+		return ret;
+	}
+
+	if ((ret = _nrf91_gpioSet(uartInfo[minor].rtspin, gpio_high)) < 0) {
+		return ret;
+	}
+
+	return EOK;
+}
+
+
+static int uart_init(unsigned int minor)
+{
+	int ret;
+	uart_t *uart;
+
+	if ((uart = uart_getInstance(minor)) == NULL) {
+		return -EINVAL;
+	}
+
+	uart->base = uartInfo[minor].base;
+	uart->tx_dma = uartInfo[minor].tx_dma;
+	uart->rx_dma = uartInfo[minor].rx_dma;
+
+	/* Disable uarte instance - changing uarte config */
+	if (*(uart->base + uarte_enable) & 0x08) {
```

Comment:
general remark - this is int -> bool cast, MISRA violation
```suggestion
	if ((*(uart->base + uarte_enable) & 0x08) != 0) {
```